File: Parser/formatter.cc
#include "formatter.hh"

using namespace Fract::Parser;

rmlx_range_result
formatter::rmlx_range(std::vector<token> *tokens) {
  rmlx_range_result _result;
  _result.index = -1;
  _result.found = false;

  std::vector<token>::iterator first; /* After iterator of open parentheses */
  std::vector<token>::iterator last; /* After iterator of close parentheses */

  /* Find open parentheses */
  for(int index = 0; index < tokens->size(); index++)
  { std::vector<token>::iterator it = tokens->begin() + index;
    if(it->type == type_open_parenthes) {
      first = ++it;
      _result.index = index;
      _result.found = true;
      break;
    }
  }

  /* Skip find close parentheses and result ready steps */
  if(!_result.found) {
    return _result;
  }

  /* Find close parentheses */
  int count = 1;
  for(int index = _result.index + 1; index < tokens->size(); index++)
  { std::vector<token>::iterator it = tokens->begin() + index;
    if(it->type == type_close_parenthes)
    { count--;
      if(count == 0)
      { last = it;
        break;
      }
    }
    else if(it->type == type_open_parenthes) {
      count++;
    }
    _result.range.push_back(*it);
  }

  /* Remove range from original tokens */
  for(int counter = 0; counter <= _result.range.size(); counter++) {
    tokens->erase(tokens->begin() + _result.index);
  }
  tokens->erase(tokens->begin() + _result.index);

  return _result;
}

File: parser/formatter/formatter.go
package formatter

import (
	"../../fract"
	"../../objects"
)

// LexRange Returns range tokens and remove from original list.
// tokens Tokens to process.
func LexRange(tokens *[]objects.Token) RangeResult {
	var (
		_result RangeResult

		first int
	)

	/* Skip find close parentheses and result ready steps */
	if !_result.Found {
		return _result
	}

	/* Find open parentheses */
	for index := 0; index < len(*tokens); index++ {
		var _token objects.Token = (*tokens)[index]
		if _token.Type == fract.TypeOpenParenthes {
			first = index
			_result.Index = index
			_result.Found = true
			break
		}
	}

	/* Find close parentheses */
	var count int = 1
	for index := _result.Index + 1; index < len(*tokens); index++ {
		var _token objects.Token = (*tokens)[index]
		if _token.Type == fract.TypeCloseParenthes {
			count = count - 1
			if count == 0 {
				break
			}
		} else if _token.Type == fract.TypeOpenParenthes {
			count = count + 1
		}
		_result.Range = append(_result.Range, _token)
	}

	/* Remove range from original tokens */
	copy((*tokens)[first:], (*tokens)[first+len(_result.Range):])
	(*tokens)[len(*tokens)-1] = *new(objects.Token)
	*tokens = (*tokens)[:len(*tokens)-1]

	return _result
}
